src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,5246 @@
+                    if (_ruleSet == AsnEncodingRules.CER && length != null)
+                    {
+                        throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                    }
+                }
+                else if (length == null)
+                {
+                    // T-REC-X.690-201508 sec 8.1.3.2 says primitive encodings must use a definite form.
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+
+                bytesRead = allBytesRead;
+                return (tag, length);
+            }
+
+            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+        }
+
+        private static void ValidateEndOfContents(Asn1Tag tag, int? length, int headerLength)
+        {
+            // T-REC-X.690-201508 sec 8.1.5 excludes the BER 8100 length form for 0.
+            if (tag.IsConstructed || length != 0 || headerLength != EndOfContentsEncodedLength)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+        }
+
+        private static ReadOnlyMemory<byte> SeekEndOfContents(
+            ReadOnlyMemory<byte> source,
+            AsnEncodingRules ruleSet,
+            int initialSliceOffset = 0)
+        {
+            ReadOnlyMemory<byte> cur = source.Slice(initialSliceOffset);
+            int totalLen = 0;
+
+            while (!cur.IsEmpty)
+            {
+                AsnReader reader = new AsnReader(cur, ruleSet);
+                (Asn1Tag tag, int? length) = reader.ReadTagAndLength(out int bytesRead);
+                ReadOnlyMemory<byte> nestedContents = reader.PeekContentBytes();
+
+                int localLen = bytesRead + nestedContents.Length;
+
+                if (tag == Asn1Tag.EndOfContents)
+                {
+                    ValidateEndOfContents(tag, length, bytesRead);
+
+                    return source.Slice(0, totalLen + initialSliceOffset);
+                }
+
+                // If the current value was an indefinite-length-encoded value
+                // then we need to skip over the EOC as well.  But we didn't want to
+                // include it as part of the content span.
+                //
+                // T-REC-X.690-201508 sec 8.1.1.1 / 8.1.1.3 indicate that the
+                // End -of-Contents octets are "after" the contents octets, not
+                // "at the end" of them.
+                if (length == null)
+                {
+                    localLen += EndOfContentsEncodedLength;
+                }
+
+                totalLen += localLen;
+                cur = cur.Slice(localLen);
+            }
+
+            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+        }
+
+        public ReadOnlyMemory<byte> PeekEncodedValue()
+        {
+            (Asn1Tag tag, int? length) = ReadTagAndLength(out int bytesRead);
```

Comment:
The signature for ReadTagAndLength seems odd -- it returns a tuple but also has an out parameter. Consider a strongly-typed return type, returning a triple or two out parameters and a return.

Reply:
Changed Tuple+out to returning the tag and issuing the two lengths as out values.